---
title: Human CD8 TILs projection
date: "`r Sys.Date()`"
author: "P. Gueguen, M. Andreatta and S. Carmona"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
  output_file=file.path(dirname(input_file), out_dir, 'Human_CD8TILs_tutorial.html'))})
---

```{r setup, echo=FALSE}
#install.packages("rmdformats")
#Template markdown setup
library(knitr)
library(rmdformats)

## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
	             cache=TRUE,
               cache.lazy=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               dev='png')
opts_knit$set(width=75)
```

**ProjecTILs** is a computational method to project scRNA-seq data into reference single-cell atlases, enabling their direct comparison in a stable, annotated system of coordinates. This tutorial outlines the main functions implemented in ProjecTILs on a small, simple dataset. For more advanced (and more biologically interesting) applications of ProjecTILs see this list of [ProjecTILs case studies](https://carmonalab.github.io/ProjecTILs_CaseStudies/)

# R environment

First, check package dependencies and install ProjecTILs

```{r message=F, warning=F}
Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS="true")
options(timeout = max(300, getOption("timeout")))

if (!requireNamespace("renv")) 
  install.packages("renv")
library(renv)
renv::restore()

remotes::install_github("carmonalab/scGate")
remotes::install_github("carmonalab/ProjecTILs")

library(ProjecTILs)
library(Seurat)
library(reshape2)
library(ggplot2)
library(gridExtra)
```

# Load reference atlas and query data

First, load the Human CD8 TILs reference atlas. This reference was built using filtered [Zheng et al.](https://www.science.org/doi/full/10.1126/science.abe6474?af=R) datasets and [semi-supervised STACAS](https://carmonalab.github.io/STACAS.demo/STACAS.demo.html#semi-supervised-integration) integration. More information about the process can be found [here](https://github.com/carmonalab/human_TIL_atlas). Included TIL subsets are literature-based. Key markers for each populations can be found below.

![](https://lh5.googleusercontent.com/gQ4sQj5qRFRaylRRQYrGs5xA6p5BWUv926aqnFju4cg9uTjbm-fpuMCYodT9xrd4FmX6L9a1hIG3hf6UfV3NGWcSMbtZO7PDFjOYOsLKEY7LfBbF50FPIQCjoCwGQ5mLhoGM29aDuhpKjLuPo1Z_kMTiXtREFK6sS8838BqRKhfOGbWWWlHzVKzHizdz)

```{r}
download.file(url = 'https://www.dropbox.com/s/phqmqlp30xgnlwn/Human_CD8_TILs_atlas.rds?dl=1', destfile = 'Human_CD8_TILs_atlas.rds')
ref <- load.reference.map(ref = 'Human_CD8_TILs_atlas.rds')
```

Let's explore the reference atlas.

```{r fig.height=4, fig.width=6}
refCols <- c('Naive' = '#b7d2e0', 'CM' = '#da6f6f','EM'= '#72b28a','TEMRA' = '#e5bfaf', 'TRM' = '#f3a166', 'TPEX' = '#aca6e0' , 'TEX' ='#f5d39f', "MAIT" = '#fdbfd4')
DimPlot(ref,label = T, cols = refCols)
```

Let's check expression of important marker genes across reference subtypes.

```{r fig.height=8, fig.width=4}
markers <- c('LEF1', "TCF7", "CCR7", "GZMK", "FGFBP2",'FCGR3A','ZNF683','ITGAE', "CRTAM",'XCL1', "HAVCR2", "TOX", "ENTPD1", 'SLC4A10','TRAV1-2')
VlnPlot(ref, features=markers, stack = T, flip = T, assay = "RNA") + NoLegend()
```

Now let's load a query dataset - [Bassez et al. 2021](https://www.nature.com/articles/s41591-021-01323-8) These are T cells from Breast cancer patients, treated or not with Immune checkpoint blockade, for which we know if they responded or not.

```{r warning=FALSE}
download.file(url = 'https://www.dropbox.com/s/dv46t7o50ldzeo5/Bassez.example.rds?dl=1', destfile = 'Bassez.example.rds')
query <- readRDS(file = 'Bassez.example.rds')
```

# Projection mode 1: aligning embeddings

The main function in ProjecTILs is `Run.ProjecTILs`, which takes as input a reference map and query dataset. The query will be batch-corrected and projected into the reference map, with low-dimensional embeddings (PCA and UMAP) compatible with those of the reference. **To minimize the batch effects, we compute filtering and projection patient per patient.**

```{r warning=FALSE,message=FALSE}
query.projected <- Run.ProjecTILs(query = query, ref = ref, filter.cell = TRUE, skip.normalize = T, split.by = 'patient_id', ncores = 4)
```

## Visualize projection

Plot projection of new data over the reference in UMAP space. The contour lines display the density of projected query cells onto the reference map.

```{r fig.height=4, fig.width=6}
plot.projection(ref, query.projected, linesize = 0.5, pointsize = 0.5, cols = refCols)
```

We can also plot the predicted composition of the query in terms of reference T cell subtypes. By default, `plot.statepred.composition` will look on the `functional.cluster` metadata that was created by the projection.

```{r fig.height=3, fig.width=5}
plot.statepred.composition(ref, query.projected,metric = "Percent", cols = refCols) + theme_bw()
```

How do the gene expression levels compare between reference and query for the different cell states?

```{r fig.height=8, fig.width=12}
plot.states.radar(ref, query=query.projected, genes4radar = markers)
```

## Compare cell states across conditions

If we have multiple conditions (e.g. control vs. treatment, or samples from different tissues), we can search for discriminant genes between conditions (otherwise, by default this analysis is performed using the reference as the 'control')

Let's start by plotting proportions between Responders and Non-responders to ICB.

```{r fig.height=8, fig.width=12}
# Splitting the Seurat object by response to ICB (called expansion here)
query.list <- subset(query.projected, subset = expansion %in% c("E",
    "NE"))
query.list <- SplitObject(query.list, split.by = "expansion")
pll <- list()
pll[[1]] <- plot.projection(ref, cols = refCols, query.list[["E"]]) + ggtitle("Responder")
pll[[2]] <- plot.statepred.composition(ref, query.list[["E"]], cols = refCols, metric = "Percent") +
    ggtitle("Responder") + ylim(0, 60)+ theme_bw()+ RotatedAxis() + NoLegend()
pll[[3]] <- plot.projection(ref, cols = refCols, query.list[["NE"]]) + ggtitle("Non-responder")
pll[[4]] <- plot.statepred.composition(ref, cols = refCols, query.list[["NE"]], metric = "Percent") +
    ggtitle("Non-responder") + ylim(0, 60) + theme_bw() + RotatedAxis() + NoLegend()
grid.arrange(grobs = pll, ncol = 2, nrow = 2, widths = c(1.5, 1))
```

We notice a increase of TEX in responders (ie, T cell expanded patients), which is consistent with the original annotation.

    'Expansion mainly involved CD8+ T cells with pronounced expression of cytotoxic-activity (PRF1, GZMB), immune-cell homing (CXCL13) and exhaustion markers (HAVCR2, LAG3)
    Conversely, undifferentiated pre-effector/memory T cells (TCF7+, GZMK+) were inversely correlated with T cell expansion.'

To make results more clear, we can check the results as fold-change of differences between responders and non-responders.

```{r}
which.types <- table(query.projected$functional.cluster) > 20  # disregard very small populations
states_all <- levels(ref$functional.cluster)
cols_use <- ref@misc$atlas.palette

norm.c <- table(query.list[["NE"]]$functional.cluster)/sum(table(query.list[["NE"]]$functional.cluster))
norm.q <- table(query.list[["E"]]$functional.cluster)/sum(table(query.list[["E"]]$functional.cluster))

foldchange <- norm.q[which.types]/norm.c[which.types]
foldchange <- sort(foldchange, decreasing = T)

tb.m <- melt(foldchange)
colnames(tb.m) <- c("Cell_state", "Fold_change")
pll <- list()
ggplot(tb.m, aes(x = Cell_state, y = Fold_change, fill = Cell_state)) + geom_bar(stat = "identity") +
    scale_fill_manual(values = cols_use) + geom_hline(yintercept = 1) + scale_y_continuous(trans = "log2") +
    theme(axis.text.x = element_blank(), legend.position = "left") + ggtitle("Responder vs. Non-responder") + theme_bw()
```

We can also check for gene expression changes between conditions.

```{r fig.height=8, fig.width=12, warning=FALSE,message=FALSE}
plot.states.radar(ref, query = query.list, min.cells = 20, genes4radar = c('LEF1', "TCF7", "CCR7", "GZMK", "FGFBP2",'FCGR3A','ZNF683','ITGAE', "CRTAM", "CD200",'GNG4', "HAVCR2", "TOX", "ENTPD1", 'SLC4A10','TRAV1-2')) 
```

# Projection mode 2: classifier

If you **do not wish to embed your query data into the reference space,** you may also simply use ProjecTILs as a cell type classifier. This may be useful to annotate cell types in your query without altering existing embeddings. This mode is quicker than mode 1.

See the query dataset in unsupervised low-dim embeddings:

```{r fig.height=4, fig.width=6}
query <- query |> FindVariableFeatures(nfeatures=500) |>
  ScaleData() |> RunPCA(npcs=10) |> RunUMAP(dims=1:10)

DimPlot(query, group.by = 'cellSubType', label = T, repel = T) 
```

The `ProjecTILs.classifier` function applies reference-projection but does not alter the current embeddings.

```{r fig.height=4, fig.width=5}
query <- ProjecTILs.classifier(query=query, ref=ref, ncores = 4)
DimPlot(query, group.by="functional.cluster", cols = refCols, label = T)
```

As compared to mode 1, **this will keep all your cells. They will just be assigned the NA label**, as they were previously removed by scGate filtering.

# Further reading

-   ðŸ—ï¸ For more reading about the CD8+ T cell atlas construction, see the [human TIL atlas repository](https://github.com/carmonalab/human_TIL_atlas)

-   ðŸ§  For applications of ProjecTILs to gain biological insight on several public datasets please see the **ProjecTILs case studies** - [INDEX](https://carmonalab.github.io/ProjecTILs_CaseStudies/) - [Repository](https://github.com/carmonalab/ProjecTILs_CaseStudies)

-   ðŸ‘‰ To generate your own reference map for ProjecTILs see [Custom reference map tutorial](https://carmonalab.github.io/ProjecTILs.demo/build_ref_atlas.html)

-   ðŸ“– ProjecTILs publication: [Andreatta et al Nat. Comm. 2021](http://dx.doi.org/10.1038/s41467-021-23324-4) and ProjecTILs [repository](https://github.com/carmonalab/ProjecTILs)
